test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -97,22 +94,90 @@
         final CompletableFuture<Void> newWritersComplete = new CompletableFuture<>();
         final CompletableFuture<Void> readersComplete = new CompletableFuture<>();
         final List<CompletableFuture<Void>> txnStatusFutureList = synchronizedList(new ArrayList<>());
-        final AtomicBoolean txnWrite = new AtomicBoolean(false);
+
+        final ConcurrentSet<UUID> committingTxn = new ConcurrentSet<>();
+        final ConcurrentSet<UUID> abortedTxn = new ConcurrentSet<>();
+        final AtomicLong eventData = new AtomicLong();
+        final boolean txnWrite;
+
+        private final ConcurrentHashMap<Long, Integer> eventMap = new ConcurrentHashMap<>();
+
+        TestState(boolean txnWrite) {
+            this.txnWrite = txnWrite;
+        }
+
+        void eventWritten(Long event) {
+            eventMap.putIfAbsent(event, 0);
+        }
+
+        void eventRead(Long event) {
+            eventMap.compute(event, (x, y) -> {
+                if (y == null) {
+                    return 1;
+                } else {
+                    return y + 1;
+                }
+            });
+        }
+
+        Long getNext() {
+            return eventData.getAndIncrement();
+        }
+
+        int getEventWrittenCount() {
+            return eventMap.size();
+        }
+
+        int getEventReadCount() {
+            return eventMap.values().stream().reduce(0, (x, y) -> x + y);
+        }
+
+        void printAnomalies() {
+            List<Long> notRead = eventMap.entrySet().stream().filter(x -> x.getValue() == 0)
+                    .map(Map.Entry::getKey).collect(Collectors.toList());
+            if (notRead.size() > 0) {
+                log.error("anomalies, unread events => {}", notRead);
+            }
+
```

Comment:
does it do any good if we print null events also ?

Reply:
what are null events?

Reply:
these ?
https://github.com/pravega/pravega/pull/2113/files#diff-7c588e31332a0f89d89d04375f736605R115

Reply:
These are not going to result in null events.. 
you cant use null as a key for a map anyway :) 
what this is doing is map.compute(key, (x,y) -> {})
this computes and updates the map based on if an entry against key is present or not. 
y will be null if map did not have an entry for this key.